In [1]:
from H2.f_pq import *

# Elementos de matriz $g_{pqrs}$

El presente archivo provee una explicación y cálculo para los elementos de matriz de dos cuerpos $g_{pqrs}$ para la molécula de hidrógeno empleando la base STO-3G. Considere que el hamiltoniano electrónico en segunda cuantización para la molécula de hidrógeno es de la forma

$$ \mathcal{H} = \sum_{pq} f_{pq} a^\dagger_p a_q + \sum_{pqrs} g_{pqrs} a^\dagger_p a^\dagger_s a_r a_q \,.$$

---

La base de funciones a considerar es
$$ \chi_1 = \psi_+ \alpha \,, \quad
\chi_2 = \psi_+ \beta \,, \quad
\chi_3 = \psi_- \alpha \,, \quad
\chi_4 = \psi_- \beta \,,
$$

donde

$$ \psi_\pm = c_\pm (\varphi_A \pm \varphi_B)
\qquad \text{y} \qquad
c_\pm = \frac{1}{\sqrt{2 (1 \pm S)}}\,, $$

con $c_\pm$ la constante de normalización y

$$ \varphi_C := R_{nl}^{k}(\zeta, \vec{d}, \vec{\alpha}, r_C) Y_{lm}(\theta,\phi) \,,$$

las funciones espaciales. En esto último $R_{nl}^{k}$ se trata de una combinación lineal de $k$ funciones Gaussianas o como también es conocido, una $k$-contracción de gaussianas:

$$ R_{nl}^{k}(\zeta, \vec{d}, \vec{\alpha}, r) = \sum_{i=1}^{k} d_i R_{l}^{\mathrm{GTO}} (\zeta^2 \alpha_i, r) \,;
\qquad \qquad
R_l^{\mathrm{GTO}} (\alpha, r) = \frac{2(2\alpha)^{3/4}}{\pi^{1/4}} \sqrt{\frac{2^l}{(2l+1)!!}} (\sqrt{2\alpha}r)^l \exp(-\alpha r^2) \,.
$$

Los elementos de matriz por calcular son de la forma

$$ g_{pqrs} = \int \frac{\chi_p^* (\mathbf{x}_1) \chi_q^* (\mathbf{x}_2) \chi_r (\mathbf{x}_1) \chi_s (\mathbf{x}_2)}{r_{12}} \mathrm{d}\vec{\mathbf{x}}_1 \, \mathrm{d}\vec{\mathbf{x}}_2 $$

Sustituyendo a las funciones base se sigue

$$\begin{align}
g_{pqrs}
    &= \int \frac{\psi_{\pm_p}^* (\mathbf{r}_1)
        \psi_{\pm_q}^* (\mathbf{r}_2)
        \psi_{\pm_r} (\mathbf{r}_1)
        \psi_{\pm_s} (\mathbf{r}_2)}{r_{12}}
        \mathrm{d}V_1 \, \mathrm{d}V_2
    \int \sigma_p^* (\omega_1) \sigma_r (\omega_1) \mathrm{d}\omega_1
    \int \sigma_q^* (\omega_2) \sigma_s (\omega_2) \mathrm{d}\omega_2
\end{align}
$$

Note que las integrales sobre las coordenadas de espín serán diferentes de 0 sí y solo sí $p$ y $r$ ($q$ y $s$) son simultáneamente par o impar, i.e. (2,4) o (1,3).

Desarrollamos para los términos iniciales de la forma $\psi_{\pm_p}^* (\mathbf{r}_1) \psi_{\pm_q}^* (\mathbf{r}_2) \psi_{\pm_r} (\mathbf{r}_1) \psi_{\pm_s} (\mathbf{r}_2)$:

$$\psi_{\pm_p}^* (\mathbf{r}_1) \psi_{\pm_q}^* (\mathbf{r}_2) \psi_{\pm_r} (\mathbf{r}_1) \psi_{\pm_s} (\mathbf{r}_2)
    = c_{\pm_p} c_{\pm_q} c_{\pm_r} c_{\pm_s} [\varphi_A (\mathbf{r}_1) \pm_p \varphi_B (\mathbf{r}_1)]^* [\varphi_A (\mathbf{r}_2) \pm_q \varphi_B (\mathbf{r}_2)]^* [\varphi_A (\mathbf{r}_1) \pm_r \varphi_B (\mathbf{r}_1)] [\varphi_A (\mathbf{r}_2) \pm_s \varphi_B (\mathbf{r}_2)]
$$

Multiplicando todos los factores del lado derecho de la igualdad se obtienen un total de 16 términos. Considere el código siguiente donde se imprime la forma de cada uno de los términos junto con el signo del mismo.

In [2]:
nuc = ['A', 'B']

term = 1
for p in nuc:
    for q in nuc:
        for r in nuc:
            for s in nuc:
                sign = ''
                if p == 'B':
                    sign += 'p'
                if q == 'B':
                    sign += 'q'
                if r == 'B':
                    sign += 'r'
                if s == 'B':
                    sign += 's'
                print(term, '-', p, q, r, s, '-' , sign,)
                term += 1

# Aquí p, q, s, r representan el signo de p, q, r ,s. Entonces, por ejemplo el término prs, no es más que el producto de los signos de p, r, s.

1 - A A A A - 
2 - A A A B - s
3 - A A B A - r
4 - A A B B - rs
5 - A B A A - q
6 - A B A B - qs
7 - A B B A - qr
8 - A B B B - qrs
9 - B A A A - p
10 - B A A B - ps
11 - B A B A - pr
12 - B A B B - prs
13 - B B A A - pq
14 - B B A B - pqs
15 - B B B A - pqr
16 - B B B B - pqrs


En general se tendrá un subtérmino de la forma

$$ \varphi_A^* (\mathbf{r}_1) \varphi_B^* (\mathbf{r}_2) \varphi_C (\mathbf{r}_1) \varphi_D (\mathbf{r}_2)
= G_{\mu}^* (\vec{r}_1, \vec{\alpha}, \vec{A}) G_{\nu}^* (\vec{r}_2, \vec{\beta}, \vec{B}) G_{\lambda} (\vec{r}_1, \vec{\gamma}, \vec{C}) G_{\sigma} (\vec{r}_2, \vec{\delta}, \vec{D})
$$

donde

$$ G_{\mu} (\vec{r}, \vec{\alpha}, \vec{A}) = \sum_{i = 1}^{k} d_{i \mu} G (\vec{r}, \alpha_{i \mu}, \vec{A}) $$

De lo cual es posible notar que la cantidad de subsubtérminos para un subtérmino arbitrario $\varphi_A (\mathbf{r}_1) \varphi_B (\mathbf{r}_2) \varphi_C (\mathbf{r}_1) \varphi_D (\mathbf{r}_2)$ son $k^4$. En nuestro caso $k=3$, i.e. un total de $3^4$ subsubtérminos.

Más explícitamente un subtérmino lucirá de la siguiente forma:

$$ \varphi_A^* (\mathbf{r}_1) \varphi_B^* (\mathbf{r}_2) \varphi_C (\mathbf{r}_1) \varphi_D (\mathbf{r}_2)
    = \sum_{i = 1}^{3} \sum_{j = 1}^{3} \sum_{k = 1}^{3} \sum_{l = 1}^{3} d_{i \mu}^* \, d_{j \nu}^* \, d_{k \lambda} \, d_{l \sigma} \,
        G^* (\vec{r}_1, \alpha_{i \mu}, \vec{A}) \, G^* (\vec{r}_2, \beta_{j \nu}, \vec{B}) \,
        G (\vec{r}_1, \gamma_{k \lambda}, \vec{C}) \,  G (\vec{r}_2, \delta_{l \sigma}, \vec{D})
$$

Introduciendo a la integral original se sigue que

$$ \int \frac{1}{r_{12}} \varphi_A^* (\mathbf{r}_1) \varphi_B^* (\mathbf{r}_2) \varphi_C (\mathbf{r}_1) \varphi_D (\mathbf{r}_2) \, \mathrm{d}V_1 \, \mathrm{d}V_2
= \sum_{i = 1}^{3} \sum_{j = 1}^{3} \sum_{k = 1}^{3} \sum_{l = 1}^{3} d_{i \mu}^* \, d_{j \nu}^* \, d_{k \lambda} \, d_{l \sigma} \,
        \mathcal{V}_{ijkl}
$$

Donde la integral $\mathcal{V}_{ijkl}$ es

$$\begin{align}
    \mathcal{V}_{ijkl}
    & = \int \frac{1}{r_{12}}
        G^* (\vec{r}_1, \alpha_{i \mu}, \vec{A}) \, G^* (\vec{r}_2, \beta_{j \nu}, \vec{B}) \,
        G (\vec{r}_1, \gamma_{k \lambda}, \vec{C}) \,  G (\vec{r}_2, \delta_{l \sigma}, \vec{D}) \,
        \mathrm{d}V_1 \, \mathrm{d}V_2 \\
    & = \left( \frac{2\alpha_{i \mu}}{\pi} \right)^{3/4} \left( \frac{2\beta_{j \nu}}{\pi} \right)^{3/4}
        \left( \frac{2\gamma_{k \lambda}}{\pi} \right)^{3/4} \left( \frac{2\delta_{l \sigma}}{\pi} \right)^{3/4}
        \underbrace{\int \frac{1}{r_{12}}
        \exp(-\alpha_{i \mu} r^2_{1A}) \, \exp(-\beta_{j \nu} r^2_{2B}) \,
        \exp(-\gamma_{k \lambda} r^2_{1C}) \, \exp(-\delta_{l \sigma} r^2_{2D}) \, \mathrm{d}V_1 \, \mathrm{d}V_2}_{(AB | CD)}
\end{align}
$$

<div class="alert alert-block alert-info">
<strong>Caso $(AB|CD)$ donde $A\neq B\neq C\neq D$.</strong>
</div>

La integral $(AB|CD)$ de acuerdo con Szabo y Ostlund [1, p.416] es:

$$ (AB|CD)
        = \frac{2 \pi^{5/2}}{(\alpha + \beta)(\gamma + \delta)(\alpha + \beta + \gamma + \delta)^{1/2}} \,
        \exp\left[ -\frac{\alpha \, \beta}{\alpha + \beta} |\mathbf{R}_A-\mathbf{R}_B|^2 -\frac{\gamma \, \delta}{\gamma + \delta} |\mathbf{R}_C-\mathbf{R}_D|^2 \right]
        F_0 \left[ \frac{(\alpha + \beta)(\gamma + \delta)}{\alpha + \beta + \gamma + \delta} |\mathbf{R}_P-\mathbf{R}_Q|^2 \right]
$$

donde $F_0(t)$ es
$$ F_0(t) = \frac{1}{2} \left( \frac{\pi}{t} \right)^{1/2} \mathrm{erf}(t^{1/2}) \,;
\quad
\text{con} \quad \mathrm{erf}(z) = \frac{2}{\sqrt{\pi}} \int_0^{z} e^{-t^2} \, \mathrm{d}t
$$

y $\mathbf{R}_P$, $\mathbf{R}_Q$ son

$$ \mathbf{R}_P = \frac{\alpha \mathbf{R}_A + \beta \, \mathbf{R}_B}{\alpha + \beta} \,,
\quad
\mathbf{R}_Q = \frac{\gamma \mathbf{R}_C + \delta \, \mathbf{R}_D}{\gamma + \delta} \,.$$

<div class="alert alert-block alert-warning">
<strong>Observación.</strong> Para Szabo y Ostlund las Gaussianas centradas en $A$, $B$ están asociadas al electrón 1, mientras que las centradas en $C$ y $D$ están asociadas al electrón 2. En nuestro caso $A$, $C$ están asociadas al electrón 1, mientras que $B$ y $D$ al electrón 2.

Por esto último, es necesario considerar las equivalencias siguientes:
- A -> A , $\alpha$ -> $\alpha$
- B -> C , $\beta$ -> $\gamma$
- C -> B , $\gamma$ -> $\beta$
- D -> D , $\delta$ -> $\delta$
</div>

In [3]:
def K2(α: float, β: float, γ: float, δ: float, RA: NDArray, RB: NDArray, RC: NDArray, RD: NDArray) -> float:
    """ Factor pre-exponencial

    Parámetros
        (α, β, γ, d) : exponente orbital Gaussiano
        (RA, RB, RC, RD) : coordenada del núcleo (A, B, C, D)
    """
    return np.exp(arg(α, γ, RA, RC) + arg(β, δ, RB, RD))

En nuestra notación se tiene entonces que
$$ (AC|BD)
        = \frac{2 \pi^{5/2}}{(\alpha + \gamma)(\beta + \delta)(\alpha + \beta + \gamma + \delta)^{1/2}} \,
        \exp\left[ -\frac{\alpha \, \delta}{\alpha + \delta} |\mathbf{R}_A-\mathbf{R}_C|^2 -\frac{\beta \, \delta}{\beta + \delta} |\mathbf{R}_B-\mathbf{R}_D|^2 \right]
        F_0 \left[ \frac{(\alpha + \gamma)(\beta + \delta)}{\alpha + \beta + \gamma + \delta} |\mathbf{R}_P-\mathbf{R}_Q|^2 \right]
$$

donde $\mathbf{R}_P$, $\mathbf{R}_Q$ serían
$$ \mathbf{R}_P = \frac{\alpha \mathbf{R}_A + \gamma \, \mathbf{R}_C}{\alpha + \gamma} \,,
\quad
\mathbf{R}_Q = \frac{\beta \, \mathbf{R}_B + \delta \, \mathbf{R}_D}{\beta + \delta} \,.$$

Observe que $(AC|BD) = (AD|BC) = (CA|DB)$, lo cual implica que de las 16 integrales iniciales por calcular solo hay 4 valores únicos.

<div class="alert alert-block alert-info">
<strong>Caso $(AA|AA)$.</strong>
</div>

La integral $(AC|BD)$ considera todos las posibles combinaciones de núcleos $A$, $B$, $C$, $D$ dado que el sistema solo tiene dos núcleos, a excepción de un caso. Cuando todos los núcleos son iguales, i.e. $A = B = C = D$, la expresión diverge puesto que $\mathbf{R}_P = \mathbf{R}_Q = \mathbf{R}_A$. En cuyo caso, la expresión a considerar se puede deducir a partir de la igualdad (A.40) en [1, 416]:

$$ (AA|AA)
    = \frac{M}{2\pi^2} \left( \frac{\pi^2}{pq} \right)^{3/2} \int k^{-2} \exp\left[-\frac{(p+q)}{4pq} k^2 \right]
        \exp\left[i \mathbf{k} \cdot (\mathbf{R}_P - \mathbf{R}_Q) \right] \, \mathrm{d}\vec{\mathbf{k}}
$$

donde $p = \alpha + \gamma$, $q = \beta + \delta$. No obstante, como $A = B = C = D$, entonces $M = 1$ y $\mathbf{R}_P - \mathbf{R}_Q = 0$. De lo cual, la integral en cuestión se reduce a

$$ (AA|AA) = \frac{2 \pi^{5/2}}{(\alpha + \gamma) (\beta + \delta) (\alpha + \beta + \gamma + \delta)^{1/2}} $$

In [4]:
def ACBD(α: float, β: float, γ: float, δ: float, RA: NDArray, RB: NDArray, RC: NDArray, RD: NDArray) -> float:
    """ Integral de dos electrones (AC|BD)
    
    (α, β, γ, δ) : exponentes orbitales Gaussianos
    (RA, RB, RC, RD) : coordenadas del núcleo (A, B, C, D)

        A, C : asociados al electrón 1
        B, D : asociados al electrón 2
    """

    factor = 2*np.power(np.pi, 5/2) / ((α + γ)*(β + δ)*np.sqrt(α + β + γ + δ))

    # caso RA=RB y RB=RC y RC=RD (i.e. 4 núcleos iguales)
    if np.array_equal(RA, RB) and np.array_equal(RB, RC) and np.array_equal(RC, RD):
        return factor
    else:
        Rp = RP(α, γ, RA, RC)
        Rq = RP(β, δ, RB, RD)
        RPQ2 = np.square(np.linalg.norm(Rp-Rq))

        # caso (α = β ^ γ = δ) o (a = γ ^ β = δ) o (a = δ ^ β = c)
        if (α == β and γ == δ) or (α == γ and β == δ) or (α == δ and β == γ):
            # caso RA=RB ^ RC=RD
            if (np.array_equal(RA, RB) and np.array_equal(RC, RD)):
                term = factor * K2(α, β, γ, δ, RA, RB, RC, RD)
            # caso RA=RD ^ RB=RC
            elif (np.array_equal(RA, RD) and np.array_equal(RB, RC)):
                term = factor * K2(α, β, γ, δ, RA, RB, RC, RD)
            # caso RA=RC ^ RB=RD
            elif (np.array_equal(RA, RC) and np.array_equal(RB, RD)):
                term = factor * F0((α + γ)*(β + δ)/(α + β + γ + δ)*RPQ2)
            else:
                term = factor * K2(α, β, γ, δ, RA, RB, RC, RD) * F0((α + γ)*(β + δ)/(α + β + γ + δ)*RPQ2)
        else:
            term = factor * K2(α, β, γ, δ, RA, RB, RC, RD) * F0((α + γ)*(β + δ)/(α + β + γ + δ)*RPQ2)

        return term

En el código la función `V12` se tratará de la integral $\mathcal{V}_{ijkl}$

$$ \mathcal{V}_{ijkl}
    = \left( \frac{2\alpha_{i \mu}}{\pi} \right)^{3/4} \left( \frac{2\beta_{j \nu}}{\pi} \right)^{3/4} \left( \frac{2\gamma_{k \lambda}}{\pi} \right)^{3/4} \left( \frac{2\delta_{l \sigma}}{\pi} \right)^{3/4} (AC | BD)
$$

In [5]:
def V12(α: float, β: float, γ: float, δ: float, RA: NDArray, RB: NDArray, RC: NDArray, RD: NDArray):
    """ Integral de dos electrones total (normalizada)
        
    (α, β, γ, δ) : exponentes orbitales Gaussianos
    (RA, RB, RC, RD) : coordenadas del núcleo (A, B, C, D)

        A, C : asociados al electrón 1
        B, D : asociados al electrón 2
    """
    v12 = GaussNorm(α) * GaussNorm(β) * GaussNorm(γ) * GaussNorm(δ) * ACBD(α, β, γ, δ, RA, RB, RC, RD)
    return v12

En el código la función `Vmn` se tratará de la suma $\mathcal{V}_{mn}$ dada por

$$ \mathcal{V}_{mn}
= \int \frac{1}{r_{12}} \varphi_A^* (\mathbf{r}_1) \varphi_B^* (\mathbf{r}_2) \varphi_C (\mathbf{r}_1) \varphi_D (\mathbf{r}_2) \, \mathrm{d}V_1 \, \mathrm{d}V_2
= \sum_{i = 1}^{3} \sum_{j = 1}^{3} \sum_{k = 1}^{3} \sum_{l = 1}^{3} d_{i \mu}^* \, d_{j \nu}^* \, d_{k \lambda} \, d_{l \sigma} \, \mathcal{V}_{ijkl}
$$

In [6]:
def Vmn(d: NDArray, a: NDArray, RA: NDArray, RB: NDArray, RC: NDArray, RD: NDArray):
    """ Elemento de matriz de interacción de dos electrones

    d : vector de coeficientes de expansión (d1, d2, ..., dk)
    a : vector de exponentes orbitales Gaussianos (α1, α2, ..., αk)
    (RA, RB, RC, RD) : coordenada del núcleo (A, B, C, D)
    """
    Mijkl = 0 # elemento de tensor
    for di, αi in zip(d, a):
        for dj, αj in zip(d, a):
            for dk, αk in zip(d, a):
                for dl, αl in zip(d, a):
                    Mijkl += di * dj * dk * dl * V12(αi, αj, αk, αl, RA, RB, RC, RD)
    
    return Mijkl

<div class="alert alert-block alert-warning">
<strong>Revisión de control.</strong>
</div>

In [7]:
# distancia interatómica de 1.4 u.a.
RA = np.array([0, 0, 0])
RB = np.array([1.4, 0, 0])

# Szabo & Ostlund
AA_AA = Vmn(d_Szabo, a_Szabo, RA, RA, RA, RA)
AB_AB = Vmn(d_Szabo, a_Szabo, RA, RB, RA, RB)
AA_AB = Vmn(d_Szabo, a_Szabo, RA, RA, RA, RB)
AA_BB = Vmn(d_Szabo, a_Szabo, RA, RA, RB, RB)
print("--- Szabo & Ostlund ---")
print('(AA|AA):', AA_AA)
print('(AB|AB):', AB_AB)
print('(AA|AB):', AA_AB)
print('(AA|BB):', AA_BB)

# Este trabajo
AA_AA = Vmn(d, a, RA, RA, RA, RA)
AB_AB = Vmn(d, a, RA, RB, RA, RB)
AA_AB = Vmn(d, a, RA, RA, RA, RB)
AA_BB = Vmn(d, a, RA, RA, RB, RB)
print("\n--- Este trabajo ---")
print('(AA|AA):', AA_AA)
print('(AB|AB):', AB_AB)
print('(AA|AB):', AA_AB)
print('(AA|BB):', AA_BB)

--- Szabo & Ostlund ---
(AA|AA): 0.7746079055149163
(AB|AB): 0.5696774985883127
(AA|AB): 0.44410895821293006
(AA|BB): 0.2997938735581481

--- Este trabajo ---
(AA|AA): 0.7746079610602591
(AB|AB): 0.5696771232827854
(AA|AB): 0.44410744195746105
(AA|BB): 0.2997953102608419


A continuación una tabla comparativa de las integrales de dos electrones entre los valores obtenidos y los de la literatura (Szabo).

|  Integral  |  $(AA|AA)$ |  $(AB|AB)$ |  $(AA|AB)$ | $(AA|BB)$ |
|:-----------|:----------:|:----------:|:----------:|:----------:|
|   Szabo    | 0.7746   | 0.5697   | 0.4441   | 0.2970   |
|   Este trabajo   | 0.7746079610602591  | 0.5696771232827854 | 0.44410744195746105 | 0.2997953102608419  |

Como aclaración, Szabo reporta los valores previos con los subíndices 1 y 2, estos corresponden (en ese mismo orden) a los núcleos $A$ y $B$. Además, para Szabo la integral $(AB|CD)$ es de tal forma que los núcleos $(A,B)$ están asociados al electrón 1 y $(C,D)$ están asociados al electrón 2. En nuestro caso los núcleos $(A,C)$ están asociados al electrón 1 y $(B,D)$ están asociados al electrón 2.

<div class="alert alert-block alert-info">
<strong>Elemento de matriz $g_{pqrs}$</strong>
</div>

In [8]:
def gpqrs(P: NDArray, d: NDArray, a: NDArray, RA: NDArray, RB: NDArray):
    """ Elemento de matriz g_pqrs

    Parámetros
        P : vector de elementos de la base, P = (p,q,r,s)
        d : vector de coeficientes de expansión (d1, d2, ..., dk)
        a : vector de exponentes orbitales Gaussianos (α1, α2, ..., αk)
        (RA, RB) : coordenada del núcleo (A, B)
    """
    p,q,r,s = P
    nucleos: list[NDArray] = [RA, RB]

    Mpqrs: float = 0.0 # elemento de tensor

    # si ocurre que (p,r) o (q,s) no son simultáneamente par o impar, g_pqrs = 0 (debido al espín)
    if (((p in [1,3] and r not in [1,3]) or (p in [2,4] and r not in [2,4]))
            or ((q in [1,3] and s not in [1,3]) or (q in [2,4] and s not in [2,4]))):
        Mpqrs = 0 # elemento de tensor
    else:
        sgn_p: int = 1 if p in [1,2] else -1 # signo de p
        sgn_q: int = 1 if q in [1,2] else -1 # signo de q
        sgn_r: int = 1 if r in [1,2] else -1 # signo de r
        sgn_s: int = 1 if s in [1,2] else -1 # signo de s
    
        # coeficientes de normalización (p,q,r,s)
        cp = cPM(d, a, RA, RB, sgn_p)
        cq = cPM(d, a, RA, RB, sgn_q)
        cr = cPM(d, a, RA, RB, sgn_r)
        cs = cPM(d, a, RA, RB, sgn_s)
        
        for A in nucleos:
            for B in nucleos:
                for C in nucleos:
                    for D in nucleos:
                        sign = 1
                        if np.array_equal(A, RB):
                            sign *= sgn_p
                        if np.array_equal(B, RB):
                            sign *= sgn_q
                        if np.array_equal(C, RB):
                            sign *= sgn_r
                        if np.array_equal(D, RB):
                            sign *= sgn_s

                        mpqrs = sign*Vmn(d, a, A, B, C, D)
                        Mpqrs += mpqrs
    
        Mpqrs = cp*cq*cr*cs*Mpqrs # multiplicar Mpqrs por coefs. de normalización
        
    return Mpqrs

In [9]:
list_P = [[1,2,1,2], [3,4,3,4], [1,3,1,3], [3,2,1,4]]

# Szabo
valores = [gpqrs(P, d_Szabo, a_Szabo, RA, RB) for P in list_P]
print("--- Szabo & Ostlund ---")
print(f'J11 = {valores[0]}')
print(f'J22 = {valores[1]}')
print(f'J12 = {valores[2]}')
print(f'K12 = {valores[3]}')

# Este trabajo
valores = [gpqrs(P, d, a, RA, RB) for P in list_P]
print("\n--- Este trabajo ---")
print(f'J11 = {valores[0]}')
print(f'J22 = {valores[1]}')
print(f'J12 = {valores[2]}')
print(f'K12 = {valores[3]}')

--- Szabo & Ostlund ---
J11 = 0.6756067100561275
J22 = 0.7212793383844265
J12 = 0.6586627074571746
K12 = 0.18125478897397107

--- Este trabajo ---
J11 = 0.6756011864877296
J22 = 0.7213418728496461
J12 = 0.6586691015495469
K12 = 0.1812577068265467


A continuación una tabla comparativa de las integrales de dos electrones entre los valores obtenidos y los reportados por Szabo & Ostlund.

|  Integral  |  $J_{11}$ |  $J_{12}$ |  $J_{22}$ | $K_{12}$ |
|:-----------|:----------:|:----------:|:----------:|:----------:|
|   Szabo    | 0.6746   | 0.6636   | 0.6975   | 0.1813   |
| Este trabajo | 0.6756011864877296 | 0.7213418728496461  | 0.6586691015495469 | 0.1812577068265467 |

La diferencia entre los valores reportados por Szabo & Ostlund y los valores obtenidos emlpleando el presente código a partir de sus coeficientes de contracción y exponentes orbitales Gaussianos se deben al error numérico acumulado.

## Corrección de precisión decimal

A continuación calculamos todos elementos de matriz diferentes de cero para $R_{AB} = 1.4 a_0$ con la finalidad de comparar sus valores numéricos y realizar algún ajuste a los valores $g_{pqrs}$.

In [10]:
elementos = set()
terminos = dict()

for p in range(1,5):
    for q in range(1,5):
        for r in range(1,5):
            for s in range(1,5):
                if p == q or r == s:
                    # la acción de los 4 operadores escalera asociados a g_iijj es 0, por lo tanto se omiten dichos elementos
                    continue
                else:
                    elemento = gpqrs([p,q,r,s], d, a, RA, RB)
                    if elemento == 0 or elemento < 1e-10:
                        # elemento es 0
                        continue
                    if elemento in elementos:
                        # si existe el valor, asocia el índice pqrs este mismo
                        terminos[f'{elemento}'].append(f'{p}{q}{r}{s}')
                    else:
                        # si no existe el valor, lo añade al conjunto y asocia el índice pqrs este mismo
                        elementos.add(elemento)
                        terminos[f'{elemento}'] = [f'{p}{q}{r}{s}']

In [11]:
for elemento in elementos:
    print(elemento, ":", terminos[f'{elemento}'])

0.6756011864877296 : ['1212', '2121']
0.6586691015495469 : ['1313', '1414', '2323', '2424']
0.18125770682654677 : ['1331', '1432', '2341', '2442']
0.18125770682654668 : ['1234', '2143']
0.1812577068265467 : ['3113', '3214', '4123', '4224']
0.6586691015495475 : ['3131', '3232', '4141', '4242']
0.18125770682654666 : ['3412', '4321']
0.7213418728496461 : ['3434', '4343']


<br>

Como las siguientes igualdades entre coeficientes se sabe ocurren:
- 1212 = 2121
- 1313 = 1414 = 2323 = 2424 = 3131 = 3232 = 4141 = 4242
- 1331 = 1432 = 2341 = 2442 = 1234 = 2134 = 3113 = 3214 = 4123 = 4224 = 3412 = 4321
- 3434 = 4343

Los valores que se reportarán de $J_{11}$, $J_{22}$, $J_{12}$ y $K_{12}$ del presente trabajo se truncarán hasta la cifra decimal no coincidente, i.e.
- $g_{1212} = J_{11} =$ 0.6756011864877296
- $g_{3434} = J_{12} =$ 0.7213418728496461
- $g_{1313} = J_{22} =$ 0.65866910154954
- $g_{1331} = K_{12} =$ 0.181257706826546